azure/services/aso/aso.go

```diff
@@ -220,11 +226,64 @@ func (r *reconciler[T]) CreateOrUpdateResource(ctx context.Context, spec azure.A
 		log.V(2).Info("resource up to date")
 		return existing, nil
 	}
-	log.V(2).Info("creating or updating resource", "diff", cmp.Diff(existing, parameters))
+	log.V(2).Info("creating or updating resource", "diff", diff)
 	return r.createOrUpdateResource(ctx, existing, parameters, resourceExists, serviceName)
 }
 
-func (r *reconciler[T]) createOrUpdateResource(ctx context.Context, existing T, parameters T, resourceExists bool, serviceName string) (T, error) {
+// PatchedParameters returns the Parameters of spec with patches applied.
+func PatchedParameters[T deepCopier[T]](ctx context.Context, scheme *runtime.Scheme, spec azure.ASOResourceSpecGetter[T], existing T) (T, error) {
+	var zero T // to be returned with non-nil errors
+	parameters, err := spec.Parameters(ctx, existing)
+	if err != nil {
+		return zero, err
+	}
+	return applyPatches(scheme, spec, parameters)
+}
+
+func applyPatches[T deepCopier[T]](scheme *runtime.Scheme, spec azure.ASOResourceSpecGetter[T], parameters T) (T, error) {
+	p, ok := spec.(Patcher)
+	if !ok {
+		return parameters, nil
+	}
+
+	var zero T // to be returned with non-nil errors
+
+	gvk, err := apiutil.GVKForObject(parameters, scheme)
+	if err != nil {
+		return zero, errors.Wrap(err, "failed to get GroupVersionKind for object")
+	}
+	parameters.SetGroupVersionKind(gvk)
+	paramData, err := json.Marshal(parameters)
+	if err != nil {
+		return zero, errors.Wrap(err, "failed to marshal JSON for patch")
+	}
+
+	for i, extraPatch := range p.ExtraPatches() {
+		jsonPatch, err := yaml.ToJSON([]byte(extraPatch))
+		if err != nil {
+			return zero, errors.Wrapf(err, "failed to convert patch at index %d to JSON", i)
```

Comment:
Can we print the patch here as well for more clarity?

Reply:
These patches could be long or contain newlines, so I'm not sure if that would help much here. I think the index gives enough info to let people know where to look.